_rules/text-node-contrast-afw4f7.md

```diff
@@ -0,0 +1,365 @@
+---
+id: afw4f7
+name: Text node has minimum contrast
+rule_type: atomic
+
+description: |
+  This rule checks that the highest possible contrast of every text node with its background meets the minimal contrast requirement
+
+accessibility_requirements:
+  wcag20:1.4.3: # Contrast (Minimum)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+  - Language
+
+authors:
+  - Brian Bors
+  - Kasper Isager
+  - Wilco Fiers
+---
+
+## Applicability
+
+Any [visible](#visible) character in a [text node](https://dom.spec.whatwg.org/#text) that is a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:
+
+- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)
+- Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled]()
+- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled]()
+
+**Note**: When the text color is the same as the background color, the element is not visible, and so is not applicable.
```

Comment:
Technically in that case, the **element** is visible, but its text node isn't.
```html
<div style="background:white"><span style="background:red; color:red">Foo</span></div>
```
The `span` is visible (it is a red rectangle, and making it transparent changes the rendered pixels). The text inside the `span` is not visible (it is red on red, making it transparent will not change anything).
This is not a problem for this rule (because it does already target individual characters in the text nodes), but this note should nonetheless be made correct.

Making the element transparent:
```html
<div style="background:white"><span style="background:red; color:red; opacity:0">Foo</span></div>
```

Making the text transparent:
```html
<div style="background:white"><span style="background:red; color:rgba(1,0,0,0)">Foo</span></div>
```